Uploader/TraceFileScanner.cs

```diff
@@ -0,0 +1,147 @@
+﻿using NLog;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+/// <summary>
+/// Scans the trace directory for trace files that are ready to upload or archive.
+/// </summary>
+public class TraceFileScanner
+{
+    private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+    private static readonly Regex TraceFileRegex = new Regex(@"^coverage_\d*_\d*.txt$");
+
+    private readonly string traceDirectory;
+    private readonly Regex versionAssemblyRegex;
+    private readonly IFileSystem fileSystem;
+
+    public TraceFileScanner(string traceDirectory, string versionAssembly, IFileSystem fileSystem)
+    {
+        this.traceDirectory = traceDirectory;
+        this.versionAssemblyRegex = new Regex(@"^Assembly=" + Regex.Escape(versionAssembly) + @".*Version:([^ ]*).*", RegexOptions.IgnoreCase);
+        this.fileSystem = fileSystem;
+    }
+
+    /// <summary>
+    /// Returns all trace files that can be uploaded or archived.
+    /// </summary>
+    public IEnumerable<ScannedFile> ListTraceFilesReadyForUpload()
+    {
+        List<string> files;
+        try
+        {
+            files = fileSystem.Directory.EnumerateFiles(traceDirectory).ToList();
+        }
+        catch (Exception e)
+        {
+            logger.Error(e, "Unable to list files in {traceDirectory}. Will retry later", traceDirectory);
+            yield break;
+        }
+        
+        logger.Debug("Scanning {fileCount} files", files.Count);
+        foreach (string filePath in files)
+        {
+            string fileName = Path.GetFileName(filePath);
+            if (!IsTraceFile(fileName))
+            {
+                logger.Debug("Skipping file that does not look like a trace file: {unknownFilePath}", filePath);
```

Comment:
i think this is uninteresting logging...

Reply:
that's why I put it on debug. In rare cases (users rename files or something like that) it could still be helpful but debug logging is disabled by default. easier to tell the user to turn on debug logging than ship them a new version of the uploader with additional log statements inserted